src/jennifer/model/base.cr

```diff
@@ -586,6 +586,23 @@ module Jennifer
         write_adapter.bulk_insert(collection)
       end
 
+      # Performs bulk import of given *collection* while ignoring models that would cause a duplicate value of any `UNIQUE` index on given *unique_fields*.
+      #
+      # Some RDBMS (like MySQL) doesn't require specifying exact constraint to be violated, therefore *unique_fields* argument by default is `[] of String`.
+      #
+      # Any callback is ignored.
+      #
+      # ```
+      # Order.create({:uid => 123})
+      # Order.upsert([
+      #   Order.new({:uid => 123}),
+      #   Order.new({:uid => 321}),
+      # ])
+      # ```
+      def self.upsert(collection : Array(self), unique_fields = [] of String)
```

Comment:
let's add override that accepts block (see second realization in `QueryBuilder::Executables#upsert`)

Reply:
Hm I'm not sure about this one. In the `QueryBuilder::Executables#upsert` variant we yield with an `ExpressionBuilder`. As far as I can see, there is no `ExpressionBuilder` involved when upserting via `Adapter::Base#upsert`?

Also if I understand correctly, the block's result is passed to the `on_conflict` argument, right? This one shouldn't be hard to implement though...

Reply:
You are correct. It is used to define `ON CONFLICT` section. `Adapter::Base#upsert` accepts hash as a last argument which allows you to pass column-condition pairs

Reply:
Yeah we can add the on_conflict hash to the API. What do you think about the ExpressionBuilder not being present when calling `Model::Base::upsert`? Can we implement the block-variant otherwise? Should it just set the on_conflict hash without passing any argument to the block?